packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -507,56 +486,66 @@ export class ProjectService {
-    }
-
-    async getAllExplores(
+    async compileProject(
         user: SessionUser,
         projectUuid: string,
-        jobUuid?: string,
-        forceRefresh: boolean = false,
-    ): Promise<(Explore | ExploreError)[]> {
-        const cachedExplores = await this.projectModel.getExploresFromCache(
+    ): Promise<{ jobUuid: string }> {
+        const job: CreateJob = {
+            jobUuid: uuidv4(),
+            jobType: JobType.COMPILE_PROJECT,
+            jobStatus: JobStatusType.STARTED,
             projectUuid,
-        );
-        if (!cachedExplores || cachedExplores.length === 0 || forceRefresh) {
-            await this.projectModel.tryWithProjectLock(
-                projectUuid,
-                jobUuid || uuidv4(),
-                this.jobModel,
-                async () => {
-                    const explores = await this.refreshAllTables(
-                        user,
-                        projectUuid,
+            steps: [{ stepType: JobStepType.COMPILING }],
+        };
+        return new Promise<{ jobUuid: string }>((resolve, reject) => {
+            const onLockFailed = async () => {
+                reject(new Error('Project is already compiling'));
+            };
+            const onLockAcquired = async () => {
+                await this.jobModel.create(job);
+                resolve({ jobUuid: job.jobUuid });
+                try {
+                    await this.jobModel.update(job.jobUuid, {
+                        jobStatus: JobStatusType.RUNNING,
+                    });
+                    const explores = await this.jobModel.tryJobStep(
+                        job.jobUuid,
+                        JobStepType.COMPILING,
+                        async () => this.refreshAllTables(user, projectUuid),
                     );
                     await this.projectModel.saveExploresToCache(
                         projectUuid,
                         explores,
                     );
-                },
-            );
-        }
-        return cachedExplores || [];
+                    await this.jobModel.update(job.jobUuid, {
+                        jobStatus: JobStatusType.DONE,
+                    });
+                } catch (e) {
+                    await this.jobModel.update(job.jobUuid, {
+                        jobStatus: JobStatusType.ERROR,
+                    });
+                }
+            };
+            this.projectModel
+                .tryWithProjectLock(projectUuid, onLockAcquired, onLockFailed)
+                .catch((e) => Logger.error(`Background job failed: ${e}`));
+        });
     }
 
     async getAllExploresSummary(
         user: SessionUser,
         projectUuid: string,
         filtered: boolean,
     ): Promise<SummaryExplore[]> {
-        const explores = await this.getAllExplores(user, projectUuid);
+        const explores = await this.projectModel.getExploresFromCache(
+            projectUuid,
+        );
+        if (!explores) {
```

Comment:
if there are no explores, just return no explores, do not trigger recompile

Reply:
I assume the update is triggered by the frontend then

Reply:
@rephus it doesn't right now - but on project creation the project is always stored in the explores cache. So outside of our PR previews / dev environment - I think it's very unlikely to hit this state

Reply:
Do we have a ticket to fix this on PR previews/dev envs asap ? I assume the e2e tests are going to fail once we merge this PR.

Reply:
Updated! I call the `projectService` in the seed file, it makes the seed file a bit more business logic-y - but it works!